module/zfs/mmp.c

```diff
@@ -0,0 +1,316 @@
+	mutex_init(&mmp->mmp_io_lock, NULL, MUTEX_DEFAULT, NULL);
+}
+
+void
+mmp_fini(dsl_pool_t *dp)
+{
+	mmp_thread_state_t *mmp = &dp->dp_mmp;
+
+	mutex_destroy(&mmp->mmp_thread_lock);
+	cv_destroy(&mmp->mmp_thread_cv);
+	mutex_destroy(&mmp->mmp_io_lock);
+	bzero(mmp, sizeof (mmp_thread_state_t));
+}
+
+static void
+mmp_thread_enter(mmp_thread_state_t *mmp)
+{
+	mutex_enter(&mmp->mmp_thread_lock);
+}
+
+void noinline
+mmp_thread_exit(mmp_thread_state_t *mmp, kthread_t **mpp)
+{
+	ASSERT(*mpp != NULL);
+	*mpp = NULL;
+	cv_broadcast(&mmp->mmp_thread_cv);
+	mutex_exit(&mmp->mmp_thread_lock);
+	thread_exit();
+}
+
+void
+mmp_thread_start(dsl_pool_t *dp)
+{
+	mmp_thread_state_t *mmp = &dp->dp_mmp;
+
+	if (spa_writeable(dp->dp_spa)) {
+		mutex_enter(&mmp->mmp_thread_lock);
+		dprintf("mmp_thread_start pool %p\n", dp);
+		mmp->mmp_last_write = gethrtime();
+		mmp->mmp_thread = thread_create(NULL, 0, mmp_thread,
+		    dp, 0, &p0, TS_RUN, defclsyspri);
+		mutex_exit(&mmp->mmp_thread_lock);
+	}
+}
+
+void noinline
+mmp_thread_stop(dsl_pool_t *dp)
+{
+	mmp_thread_state_t *mmp;
+
+	ASSERT(dp);
+	ASSERT(dp->dp_mmp.mmp_thread);
+
+	mmp = &dp->dp_mmp;
+
+	dprintf("mmp_thread_stop signalling exit: pool %p\n", dp);
+
+	mutex_enter(&mmp->mmp_thread_lock);
+	mmp->mmp_thread_exiting = 1;
+	cv_broadcast(&mmp->mmp_thread_cv);
+
+	while (mmp->mmp_thread) {
+		dprintf("mmp_thread_stop waiting for exit: pool %p\n", dp);
+		cv_wait(&mmp->mmp_thread_cv, &mmp->mmp_thread_lock);
+	}
+	mutex_exit(&mmp->mmp_thread_lock);
+
+	ASSERT(mmp->mmp_thread == NULL);
+	dprintf("mmp_thread_stop completed exit: pool %p\n", dp);
+	mmp->mmp_thread_exiting = 0;
+}
+
+/*
+ * Randomly choose a leaf vdev, to write an MMP block to.  It must be
+ * writable.  It must not have an outstanding mmp write (if so then
+ * there is a problem, and a new write will also block).
+ *
+ * We try 10 times to pick a random leaf without an outstanding write.
+ * If 90% of the leaves have pending writes, this gives us a >65%
+ * chance of finding one we can write to.  There will be at least
+ * (zfs_mmp_fail_intervals) tries before the inability to write an MMP
+ * block causes serious problems.
+ */
+
+vdev_t *
+vdev_random_leaf(spa_t *spa)
+{
+	vdev_t *vd, *child;
+	int pending_writes = 10;
+
+	ASSERT(spa);
+	ASSERT(spa_config_held(spa, SCL_STATE, RW_READER) == SCL_STATE);
+
+	/*
+	 * Since we hold SCL_STATE, neither pool nor vdev state can
+	 * change.  Therefore, if the root is not dead, there is a
+	 * child that is not dead, and so on down to a leaf.
+	 */
+
+	if (vdev_is_dead(spa->spa_root_vdev))
+		return (NULL);
+
+	vd = spa->spa_root_vdev;
+	while (!vd->vdev_ops->vdev_op_leaf) {
+		child = vd->vdev_child[spa_get_random(vd->vdev_children)];
+
+		if (vdev_is_dead(child))
+			continue;
+
+		if (child->vdev_ops->vdev_op_leaf &&
+		    child->vdev_mmp_pending) {
+			if (pending_writes-- > 0)
+				continue;
+			else
+				return (NULL);
+		}
+
+		vd = child;
+	}
+	return (vd);
+}
+
+static void
```

Comment:
I would agree.  Even `mmp_delay = (new_delay + 15 * mmp_delay) / 16` or higher, since this is decaying every mmp_interval (on the order of seconds), so we want it to decay more slowly (i.e. over tens of seconds or more).